src/Control/Algebra.hs

```diff
@@ -78,57 +116,61 @@ run = runIdentity
 -- 1. Due to [a problem with recursive type families](https://gitlab.haskell.org/ghc/ghc/issues/8095), this can lead to significantly slower compiles.
 --
 -- 2. It defeats @ghc@’s warnings for redundant constraints, and thus can lead to a proliferation of redundant constraints as code is changed.
+--
+-- @since 1.0.0.0
 type Has eff sig m = (Members eff sig, Algebra sig m)
 
 -- | Construct a request for an effect to be interpreted by some handler later on.
+--
+-- @since 0.1.0.0
 send :: (Member eff sig, Algebra sig m) => eff m a -> m a
-send = alg id . inj
+send sig = runIdentity <$> alg (fmap Identity . runIdentity) (inj sig) (Identity ())
 {-# INLINE send #-}
 
 
 -- base
 
 instance Algebra (Lift IO) IO where
-  alg hom (LiftWith with k) = with (Identity ()) (coerce . hom . runIdentity) >>= hom . k . runIdentity
+  alg hdl (LiftWith with k) ctx = with hdl ctx >>= hdl . fmap k
 
 instance Algebra (Lift Identity) Identity where
-  alg hom (LiftWith with k) = with (Identity ()) (coerce . hom . runIdentity) >>= hom . k . runIdentity
+  alg hdl (LiftWith with k) ctx = with hdl ctx >>= hdl . fmap k
 
 instance Algebra Choose NonEmpty where
-  alg hom (Choose m) = hom (m True) S.<> hom (m False)
+  alg hdl (Choose m) ctx = hdl (m True <$ ctx) S.<> hdl (m False <$ ctx)
 
 instance Algebra Empty Maybe where
-  alg _ Empty = Nothing
+  alg _ Empty _ = Nothing
 
 instance Algebra (Error e) (Either e) where
-  alg hom = \case
+  alg hdl sig ctx = case sig of
     L (Throw e)     -> Left e
-    R (Catch m h k) -> either (hom . k <=< hom . h) (hom . k) (hom m)
+    R (Catch m h k) -> either (hdl . (<$ ctx) . h) pure (hdl (m <$ ctx)) >>= hdl . fmap k
 
 instance Algebra (Reader r) ((->) r) where
-  alg hom = \case
-    Ask       k -> \ r -> hom (k r) r
-    Local f m k -> \ r -> hom (k (hom m (f r))) r
+  alg hdl sig ctx = case sig of
+    Ask       k -> \ r -> hdl (k r <$ ctx) r
+    Local f m k -> \ r -> hdl (fmap k (hdl (m <$ ctx) (f r))) r
```

Comment:
We lose `LambdaCase` with the new parameter ordering, but it could be recovered by writing this tacitly in the context functor, e.g. writing the initial action as a continuation from `ctx ()` and then composing continuations with `>=>`. I’ve opted not to do that here because it doesn’t make the instances any clearer.